File: Advent of Code 2020/day2.py
old_policy = 0
new_policy = 0

for line in open("inputs/day2.in"):
    a, b, password = line.split()
    l, h = a.split("-")
    l = int(l)
    h = int(h)
    c = b[0]
    if l <= password.count(c) <= h:
        old_policy += 1
    if (password[l - 1] + password[h - 1]).count(c) == 1:
        new_policy += 1

print("Part 1:", old_policy)
print("Part 2:", new_policy)

File: Advent of Code 2021/lib/day2.ex
defmodule Day2 do
end

{:ok, file} = File.read("inputs/day2.in")

instructions = String.split(file, "\n", trim: true)

coords =
  instructions
  |> Enum.reduce({0, 0}, fn x, acc ->
    # acc = {x, y}
    [direction, n] = String.split(x)

    case direction do
      "forward" ->
        {elem(acc, 0) + String.to_integer(n), elem(acc, 1)}

      "up" ->
        {elem(acc, 0), elem(acc, 1) - String.to_integer(n)}

      "down" ->
        {elem(acc, 0), elem(acc, 1) + String.to_integer(n)}
    end
  end)

IO.inspect(elem(coords, 0) * elem(coords, 1), label: "Part 1")

coords =
  instructions
  |> Enum.reduce({0, 0, 0}, fn x, acc ->
    # acc = {x, y, aim}
    [direction, n] = String.split(x)

    case direction do
      "forward" ->
        {elem(acc, 0) + String.to_integer(n), elem(acc, 1) + String.to_integer(n) * elem(acc, 2),
         elem(acc, 2)}

      "up" ->
        {elem(acc, 0), elem(acc, 1), elem(acc, 2) - String.to_integer(n)}

      "down" ->
        {elem(acc, 0), elem(acc, 1), elem(acc, 2) + String.to_integer(n)}
    end
  end)

IO.inspect(elem(coords, 0) * elem(coords, 1), label: "Part 2")
